### PR TITLE
Handle of invalid datetime when format check fails

### PR DESCRIPTION
**Pull request by hikinggrass**
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: lib/ocpp/common/types.cpp [hunk 1]
```diff
@@ -53,8 +43,7 @@ void DateTimeImpl::from_rfc3339(const std::string& timepoint_str) {
             in.seekg(0);
             in >> date::parse("%FT%T", this->timepoint);
             if (in.fail()) {
-                EVLOG_error << "Timepoint string parsing failed. Could not convert: \"" << timepoint_str
-                            << "\" into DateTime.";
+                throw TimePointParseException(timepoint_str);
```
Comment (maaikez) [inline lib/ocpp/common/types.cpp:46]: Hmm now you added a throw where the software did not crash in the past. Did you check everywhere this is used and provide a try / catch around it if needed?

Comment (jannejy) [reply lib/ocpp/common/types.cpp:46]: Well, the `std::string` constructor calls only when json is converted to the `DateTime` type or in database conversions. In other places the constructor is not (should not be) called. So then it should be caught either by `TimePointParseException` or `std::exception`, that is done. From my side and for 2.0.1, I don't see the missing places. @hikinggrass can approve as well for 1.6

File: include/ocpp/common/types.hpp [hunk 2]
```diff
@@ -33,6 +33,23 @@ struct Message {
...
 };
 
+/// \brief Exception used when DateTime class is initialized by invalid timepoint string.
+class TimePointParseException : public std::exception {
+public:
+    explicit TimePointParseException(const std::string& timepoint_str) :
+        msg{"Timepoint string parsing failed. Could not convert: \"" + timepoint_str + "\" into DateTime."} {
+    }
+
+    ~TimePointParseException() noexcept override = default;
```
Comment (marcemmers) [inline include/ocpp/common/types.hpp:43]: Don't think we need to explicitly define this if it is a default destructor

Comment (jannejy) [reply include/ocpp/common/types.hpp]: Just copied from other exception classes here, but indeed, the compiler will generate the same destructor for us

**Review by maaikez — APPROVED** [review]

**Pull request merged**